File: haushaltsbuch/src/main/java/org/github/haschi/haushaltsbuch/api/_Währungsbetrag.java
package org.github.haschi.haushaltsbuch.api;

import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import org.apache.commons.lang3.StringUtils;
import org.github.haschi.haushaltsbuch.infrastruktur.modellierung.de.Eingehüllt;
import org.github.haschi.haushaltsbuch.infrastruktur.modellierung.de.Umhüller;
import org.immutables.value.Value;
import org.javamoney.moneta.Money;

import javax.money.Monetary;
import javax.money.MonetaryAmount;
import javax.money.format.MonetaryAmountFormat;
import javax.money.format.MonetaryFormats;
import java.util.Locale;

@JsonDeserialize(using = WährungsbetragDeserialisierer.class, as = Währungsbetrag.class)
@JsonSerialize(using = WährungsbetragSerialisierer.class)
@Eingehüllt
@Value.Immutable
public abstract class _Währungsbetrag extends Umhüller<MonetaryAmount>
{
    public static Währungsbetrag währungsbetrag(final String betrag)
    {
        if (StringUtils.isEmpty(betrag))
        {
            throw new IllegalArgumentException("Währungsbetrag ist leer");
        }

        final DeutschenWährungsbetragAnalysieren analysieren = new DeutschenWährungsbetragAnalysieren();
        final MonetaryAmount s = analysieren.aus(betrag);

        return Währungsbetrag.of(s);
    }

    public static Währungsbetrag NullEuro()
    {
        return Währungsbetrag.of(
                Money.of(0, Monetary.getCurrency(Locale.GERMANY)));
    }

    @Override
    // @JsonProperty("betrag")
    public final String toString()
    {
        if (wert() != null )
        {
            final MonetaryAmountFormat deutschesFormat = MonetaryFormats.getAmountFormat(Locale.GERMANY);
            return deutschesFormat.format(wert());
        } else {
            return StringUtils.EMPTY;
        }
    }
}

File: haushaltsbuch/src/main/kotlin/org/github/haschi/haushaltsbuch/api/Währungsbetrag.kt
package org.github.haschi.haushaltsbuch.api

import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import org.javamoney.moneta.Money
import java.util.*
import javax.money.Monetary
import javax.money.MonetaryAmount
import javax.money.format.MonetaryFormats

@JsonDeserialize(using = WährungsbetragDeserialisierer::class)
@JsonSerialize(using = WährungsbetragSerialisierer::class)
data class Währungsbetrag(val wert: MonetaryAmount)
{

    override // @JsonProperty("betrag")
    fun toString(): String
    {
        val deutschesFormat = MonetaryFormats.getAmountFormat(Locale.GERMANY)
        return deutschesFormat.format(wert)
    }

    companion object
    {
        fun währungsbetrag(betrag: String): Währungsbetrag
        {
            if (betrag.isEmpty())
            {
                throw IllegalArgumentException("Währungsbetrag ist leer")
            }

            val analysieren = DeutschenWährungsbetragAnalysieren()
            val s = analysieren.aus(betrag)

            return Währungsbetrag(s)
        }

        val NullEuro: Währungsbetrag =
                Währungsbetrag(
                        Money.of(
                                0,
                                Monetary.getCurrency(Locale.GERMANY)))

        fun euro(betrag: Double): Währungsbetrag
        {
            return Währungsbetrag(Money.of(betrag, Monetary.getCurrency(Locale.GERMANY)));
        }
    }

    operator fun plus(y: Währungsbetrag): Währungsbetrag
    {
        return Währungsbetrag(this.wert.add(y.wert))
    }
}
